src/metabase/lib/query.cljc

```diff
@@ -8,11 +8,68 @@
    [metabase.lib.metadata.calculation :as lib.metadata.calculation]
    [metabase.lib.normalize :as lib.normalize]
    [metabase.lib.options :as lib.options]
+   [metabase.lib.ref :as lib.ref]
    [metabase.lib.schema :as lib.schema]
    [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.lib.stage :as lib.stage]
    [metabase.lib.util :as lib.util]
+   [metabase.mbql.util.match :as mbql.match]
    [metabase.util.malli :as mu]))
 
+(defn- find-clause
+  [query stage-number target-clause]
+  (let [stage (lib.util/query-stage query stage-number)
+        [target-type _opts target-id] target-clause]
+    (not-empty
+      (reduce
+        (fn [accum top-level-clause]
+          (if (mbql.match/match-one (get stage top-level-clause)
+                [target-type _ target-id] true)
+            (conj accum top-level-clause)
+            accum))
+        #{}
+        [:order-by :aggregation :breakout :filter :expressions :joins :fields]))))
+
+(defn- check-subsequent-stages!
+  "Throws if target-clause is used in a subsequent stage"
+  [previous-query query stage-number target-clause]
+  (let [[_ _ target-id] target-clause]
+    (loop [stage-number stage-number]
+      (when-let [next-stage (lib.util/next-stage-number query stage-number)]
+        (when-not (->> (lib.stage/visible-columns query next-stage)
+                       (some (fn [{:keys [lib/source id] :as _column}]
+                               (and (= :source/previous-stage source) (= target-id id)))))
```

Comment:
I find this logic hard to follow. Isn't it enough to check if the removed clause is referenced in subsequent stages? Can it be visible even though it has been removed?

Reply:
Added a comment, but I think in the case of the final `breakout` or `fields` being removed, it's generally safe and not surprising to keep using the same reference transparently.